packages/big/int256/int256.gno

```diff
@@ -0,0 +1,557 @@
+	}
+	return 1
+}
+
+func New() *Int {
+	return &Int{
+		abs: new(uint256.Uint),
+	}
+}
+
+// SetInt64 sets z to x and returns z.
+func (z *Int) SetInt64(x int64) *Int {
+	neg := false
+	if x < 0 {
+		neg = true
+		x = -x
+	}
+	if z.abs == nil {
+		panic("abs is nil")
+	}
+	z.abs = z.abs.SetUint64(uint64(x))
+	z.neg = neg
+	return z
+}
+
+// SetUint64 sets z to x and returns z.
+func (z *Int) SetUint64(x uint64) *Int {
+	if z.abs == nil {
+		panic("abs is nil")
+	}
+	z.abs = z.abs.SetUint64(x)
+	z.neg = false
+	return z
+}
+
+// NewInt allocates and returns a new Int set to x.
+func NewInt(x int64) *Int {
+	return New().SetInt64(x)
+}
+
+func FromDecimal(s string) (*Int, error) {
+	return new(Int).SetString(s)
+}
+
+func MustFromDecimal(s string) *Int {
+	z, err := FromDecimal(s)
+	if err != nil {
+		panic(err)
+	}
+	return z
+}
+
+// SetString sets s to the value of z and returns z and a boolean indicating success.
+func (z *Int) SetString(s string) (*Int, error) {
+	origin := s
+	neg := false
+	// Remove max one leading +
+	if len(s) > 0 && s[0] == '+' {
+		neg = false
+		s = s[1:]
+	}
+
+	if len(s) > 0 && s[0] == '-' {
+		neg = true
+		s = s[1:]
+	}
+	var (
+		abs *uint256.Uint
+		err error
+	)
+	abs, err = uint256.FromDecimal(s)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Int{
+		abs,
+		neg,
+	}, nil
+}
+
+// // setFromScanner implements SetString given an io.ByteScanner.
+// // For documentation see comments of SetString.
+// func (z *Int) setFromScanner(r io.ByteScanner, base int) (*Int, bool) {
+// 	if _, _, err := z.scan(r, base); err != nil {
+// 		return nil, false
+// 	}
+// 	// entire content must have been consumed
+// 	if _, err := r.ReadByte(); err != io.EOF {
+// 		return nil, false
+// 	}
+// 	return z, true // err == io.EOF => scan consumed all content of r
+// }
+
+func (z *Int) Add(x, y *Int) *Int {
+	neg := x.neg
+
+	if x.neg == y.neg {
+		// x + y == x + y
+		// (-x) + (-y) == -(x + y)
+		z.abs = z.abs.Add(x.abs, y.abs)
+	} else {
+		// x + (-y) == x - y == -(y - x)
+		// (-x) + y == y - x == -(x - y)
+		if x.abs.Cmp(y.abs) >= 0 {
+			z.abs = z.abs.Sub(x.abs, y.abs)
+		} else {
+			neg = !neg
+			z.abs = z.abs.Sub(y.abs, x.abs)
+		}
+	}
+	z.neg = neg // 0 has no sign
+	return z
+}
+
+// Sub sets z to the difference x-y and returns z.
+func (z *Int) Sub(x, y *Int) *Int {
+	neg := x.neg
+	if x.neg != y.neg {
+		// x - (-y) == x + y
+		// (-x) - y == -(x + y)
+		z.abs = z.abs.Add(x.abs, y.abs)
+	} else {
+		// x - y == x - y == -(y - x)
+		// (-x) - (-y) == y - x == -(x - y)
+		if x.abs.Cmp(y.abs) >= 0 {
+			z.abs = z.abs.Sub(x.abs, y.abs)
+		} else {
+			neg = !neg
+			z.abs = z.abs.Sub(y.abs, x.abs)
+		}
+	}
+	z.neg = neg // 0 has no sign
+	return z
+}
+
+// Mul sets z to the product x*y and returns z.
+func (z *Int) Mul(x, y *Int) *Int {
+	z.abs = z.abs.Mul(x.abs, y.abs)
+	z.neg = x.neg != y.neg // 0 has no sign
+	return z
+}
+
+// Rsh sets z = x >> n and returns z.
+func (z *Int) Rsh(x *Int, n uint) *Int {
```

Comment:
In EVM, int256s are represented by signed 256bit integers, whereas int256.Int is using 257bit one. I am curious if shifting operation(and others too) might create any possible out-of-256bit range values. We can either add a range checking logic on operations, or we can first merge this and see if there is any bugs happening on the testnet.

Reply:
I prefer latter. AFAIK solidity <= 0.5.0 doesn't care about overflow/underflow and that's why openzeppelin provide SafeMath